lib/writers/git.go

```diff
@@ -68,125 +68,139 @@ func NewGitWriter(logger logr.Logger, stateStoreSpec platformv1alpha1.GitStateSt
+		logger := logger.WithValues(
+			"filepath", worktreeFilePath,
+		)
+
+		///tmp/git-dir/worker-cluster/resources/<rr-namespace>/<promise-name>/<rr-name>/foo/bar/baz.yaml
+		absoluteFilePath := filepath.Join(localTmpDir, worktreeFilePath)
+
+		//We need to protect against paths containg `..`
+		//filepath.Join expands any '../' in the path to the actual, e.g. /tmp/foo/../ resolves to /tmp/
+		//To ensure they can't write to files on disk outside of the tmp git repostiroy we check the absolute path
+		//returned by `filepath.Join` is still contained with the git repository:
+		// Note: This means `../` can still be used, but only if the end result is still contained within the git repository
+		if !strings.HasPrefix(absoluteFilePath, localTmpDir) {
+			logger.Error(nil, "path of file to write is not located within the git repostiory", "absolutePath", absoluteFilePath, "tmpDir", localTmpDir, "repoPath", worktreeFilePath, "path", g.path)
+			return nil //We don't want to retry as this isn't a recoverable error. Log error and return nil.
+		}
 
-	return nil
-}
+		if os.MkdirAll(filepath.Dir(absoluteFilePath), 0700); err != nil {
+			logger.Error(err, "could not generate local directories")
+			return err
+		}
 
-func (g *GitWriter) RemoveObject(fileName string) error {
-	log := g.Log.WithValues("dir", g.path, "fileName", fileName)
+		if err := os.WriteFile(absoluteFilePath, []byte(item.Content), 0644); err != nil {
+			logger.Error(err, "could not write to file")
+			return err
+		}
 
-	repoPath, err := createLocalDirectory()
-	if err != nil {
-		log.Error(err, "could not create temporary repository directory")
-		return err
+		if _, err := worktree.Add(worktreeFilePath); err != nil {
+			logger.Error(err, "could not add file to worktree")
+			return err
+		}
+		filesCommitted = append(filesCommitted, worktreeFilePath)
 	}
-	defer os.RemoveAll(filepath.Dir(repoPath))
 
-	repo, err := g.cloneRepo(repoPath)
-	if err != nil {
-		log.Error(err, "could not clone repository")
-		return err
-	}
+	return g.commitAndPush(repo, worktree, Add, filesCommitted, logger)
+}
 
-	worktree, err := repo.Worktree()
+func (g *GitWriter) RemoveObject(filePath string) error {
+	logger := g.Log.WithValues("dir", g.path, "filepath", filePath)
+
+	localTmpDir, repo, worktree, err := g.setupLocalDirectoryWithRepo(logger)
 	if err != nil {
-		log.Error(err, "could not access repo worktree")
 		return err
 	}
+	defer os.RemoveAll(filepath.Dir(localTmpDir))
 
-	objectFileName := filepath.Join(g.path, fileName)
-	if _, err := worktree.Filesystem.Lstat(objectFileName); err == nil {
-		if _, err := worktree.Remove(objectFileName); err != nil {
-			log.Error(err, "could not remove file from worktree")
+	worktreeFilepath := filepath.Join(g.path, filePath)
+	if _, err := worktree.Filesystem.Lstat(worktreeFilepath); err == nil {
+		if _, err := worktree.Remove(worktreeFilepath); err != nil {
+			logger.Error(err, "could not remove file from worktree")
```

Comment:
`worktree.Remove` implementation calls `Lstat` prior to deleting, and return the error if `Lstat` errors. I'm assuming that the error is not useful for a check?

Reply:
ohh interesting I didn't know that. I think its useful as is to distinguish if the error from `remove` is because it doesn't exist vs something else